### PR TITLE
rule policy documentation

### PR DESCRIPTION
**Pull request by wochinge**
**Proposed changes**:
- fix https://github.com/RasaHQ/rasa/issues/6184

Open Todos
- [x] add rules to glossary
- [x] finish migration guide or decide to wait for migration
- [x] add links from stories to rules
- [x] describe how to submit form via rule

**Required code todos** (things which are described in the docs, but don't work yet as they are not yet implemented):
- [x] Core fallback for rules https://github.com/RasaHQ/rasa/issues/6208
- [x] re-adding `ambiguity_threshold` https://github.com/RasaHQ/rasa/issues/6244
- [ ] custom actions to ask for slot https://github.com/RasaHQ/rasa/issues/6185

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by wochinge** [comment]
@melindaloubser1 @b-quachtran : Ella said that one of you would probably best to help me writing the documentation for the RulePolicy. Could one of you have maybe have a look at the overall structure? 🙏  Happy to jump on a call as well . 
My plan would be:

1. Verify overall structure with you and iterate on it.
2. Add content and iterate on it with you
3. Merge it
4. Probably do another iteration as the training data and everything is changing at the moment 😆 

I am on vacation for a week starting July 23rd, so ideally we'd get it to state where somebody can take it over while I'm away .

**Comment by b-quachtran** [comment]
@wochinge Yes! I'll take a look and let you know if I have any questions.

File: docs/core/rules.rst [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    can't handle yet. Unless these messages are close to existing training data, Rasa
+    NLU will typically show low intent confidence levels for these messages. You can
+    use :ref:`fallback-classifier` to treat low NLU confidence like an FAQ
+* Forms (TODO: LINK): It's a common use case for assistants to collect form-like data
+    from the user. Both, activation of forms as well as handling of unexpected
+    events as part of a form, are usually following fixed paths.
+
+    (OTHER ideas?)
+
+Do we need a rules vs. stories section?
```
Comment (b-quachtran) [inline docs/core/rules.rst:29]: Rules vs. stories would definitely be helpful to clarify when to use one vs. the other and how they take precedent

**Comment by b-quachtran** [comment]
It might make sense to put FAQ and Fallback into separate pages that the Rules doc can introduce & reference (like w/ Forms), since customers frequently ask about how to implement them. @melindaloubser1 thoughts?

**Comment by wochinge** [comment]
> It might make sense to put FAQ and Fallback into separate pages that the Rules doc can introduce & reference (like w/ Forms), since customers frequently ask about how to implement them. @melindaloubser1 thoughts?

Fallback already got its own page, so make sense to keep it separate. 
Not so sure about the FAQs to be honest. However, I think it's not a huge deal breaker as we can quickly copy that over to a new page then 👍

File: docs/core/forms.rst [hunk 2]
```diff
@@ -27,340 +27,56 @@ Rasa Core.
...
-configuration file. For example:
-
-.. code-block:: yaml
-
-  policies:
-    - name: "FormPolicy"
-
-see ``examples/formbot/config.yml`` for an example.
-
-Form Basics
```
[the remark this answers is not in the document]
Comment (wochinge) [reply docs/core/forms.rst:54]: yep 👍

[the remark this answers is not in the document]
Comment (wochinge) [reply docs/core/forms.rst:54]: @b-quachtran I'd rephrase / rewrite the forms stuff according to our changes and then simply add a link to the previous docs for the deprecated `FormPolicy`?

**Comment by melindaloubser1** [comment]
> > It might make sense to put FAQ and Fallback into separate pages that the Rules doc can introduce & reference (like w/ Forms), since customers frequently ask about how to implement them. @melindaloubser1 thoughts?> 


I'm guessing it'll be a short piece - rules for FAQ pretty much always look the same right? So ja, I don't think it matters much either way

File: docs/core/stories.rst [hunk 3]
```diff
@@ -3,6 +3,8 @@
 
 .. _stories:
 
+TODO: Place clever links to rules where it makes sense?
```
Comment (b-quachtran) [inline docs/core/stories.rst:6]: Hmm, also thinking an example of how to switch between stories & rules would be helpful (e.g. how to handle an FAQ intent/response in the middle of a story)

Comment (wochinge) [reply docs/core/stories.rst:6]: what do you mean by switching? The policies will give their predictions individually and the one with the highest confidence will be chosen. If your ML policy is trained for the FAQ question, then it'll either match the RulePolicy confidence (unlikely but I think theoretically possible) or it will have a (slightly) less confident prediction and the RulePolicy wins

Comment (b-quachtran) [reply docs/core/stories.rst:6]: I'm thinking of scenarios where the user switches to an FAQ question in the middle of a story and follow-up prompt is given to take the user back into the original story. It's something I ran into while porting the Sara bot and, from what I understand, the rule policy will have to handle the follow-up prompt but it wasn't super clear where the delineation exists between the rule & story.

Comment (b-quachtran) [reply docs/core/stories.rst:6]: Happy to jump on a call to explain further

**Comment by wochinge** [comment]
@b-quachtran Added docs for forms (except the part about unhappy paths) and fallbacks. I didn't proof read it at all so far, but I'd highly appreciate feedback on a higher level (everything covered, does the structure make sense etc.)

**Comment by wochinge** [comment]
Ready for review @b-quachtran 🎉

File: docs/core/fallback-actions.rst [hunk 4]
```diff
@@ -8,87 +8,74 @@ Fallback Actions
...
 
-  policies:
-    - name: "FallbackPolicy"
-      nlu_threshold: 0.4
-      core_threshold: 0.3
-      ambiguity_threshold: 0.1
-      fallback_action_name: "action_default_fallback"
+    - rule: Ask the user to rephrase whenever they send a message with low NLU confidence
+      steps:
+      - ...
```
Comment (Ghostvv) [inline docs/core/fallback-actions.rst:42]: this will be updated later?

File: docs/core/fallback-actions.rst [hunk 5]
```diff
@@ -99,30 +86,74 @@ the same manner as the ``FallbackPolicy``).
...
+the threshold for this as follows:
+
+.. code-block:: yaml
+
+  policies:
+    - name: RulePolicy
+      # Confidence threshold for the `fallback_action_name` to apply.
+      # The action will apply if no other action was predicted with
+      # a confidence >= core_fallback_threshold
+      core_fallback_threshold: 0.4
```
Comment (Ghostvv) [inline docs/core/fallback-actions.rst:144]: this is not currently implemented right?

File: docs/core/forms.rst [hunk 6]
```diff
@@ -8,278 +8,404 @@ Forms
...
+
+            return [SlotSet("user_message_length", len(text_of_last_user_message))]
+
+
+Requesting Extra Slots
+~~~~~~~~~~~~~~~~~~~~~~
+
+If you have frequent changes to the required slots and don't want to retrain your
+assistant when something changes, you can also use a
+:ref:`custom action<custom-actions>` ``action_validate_{form_name}`` to define
```
Comment (Ghostvv) [inline docs/core/forms.rst:369]: did we change it to `action_validate_`? currently in the code it is `validate_`

File: docs/core/forms.rst [hunk 7]
```diff
@@ -8,278 +8,404 @@ Forms
...
+which slots should be requested. Rasa Open Source will run your custom action whenever
+the form is executed. Set the slot ``requested_slot`` to the name of the slot which
```
Comment (Ghostvv) [inline docs/core/forms.rst:371]:
```suggestion
the form is validates user input. Set the slot ``requested_slot`` to the name of the slot which
```

File: docs/core/forms.rst [hunk 8]
```diff
@@ -8,278 +8,404 @@ Forms
...
+should be extracted next. If all desired slots are filled, set ``requested_slot``
+to ``None``.
+
+The following example shows the implementation of a custom action which requests
+the three slots ``last_name``, ``first_name``, and ``city``.
+
```
Comment (Ghostvv) [inline docs/core/forms.rst:377]: this example might be confusing, because it repeats the default behavior. @akelad what do you think?

File: docs/core/forms.rst [hunk 9]
```diff
@@ -321,35 +447,33 @@ learning to build these stories.
...
-ask if they are looking for somewhere with outside seating.
+Using a Custom Action to Ask For the Next Slot
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-You can achieve this by writing some logic into the ``required_slots()`` method,
-for example:
+As soon as the form determined which slot has to be filled next by the user, it will
+execute the action ``utter_ask_{slot_name}`` to ask the user to provide the necessary
+input. If a regular utterance is not enough, you can also provide a custom action
+``action_ask_{slot_name}`` to ask for the next slot.
```
Comment (Ghostvv) [inline docs/core/forms.rst:456]: this is not supported yet

File: docs/core/rules.rst [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+Rules
+=====
+
+.. edit-link::
+
+.. contents::
+   :local:
+
+Rasa rules are a form of training data used to train Rasa's dialogue management models.
```
Comment (Ghostvv) [inline docs/core/rules.rst:14]: "rules are a form" might be confusing, since the form exist in rasa on it's own

File: docs/core/rules.rst [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+      - action: utter_greet
+
+.. _rules-applying-anytime:
+
+Rules which Apply Anytime
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To indicate that a rule can apply at any point in a conversation, specify ``...`` as
+first conversation step. The ``...`` indicate that the rule applies independent of the
+previous conversation.
```
Comment (Ghostvv) [inline docs/core/rules.rst:72]: this will completely change. Would it make sense to write the docs for the new version?

File: docs/core/rules.rst [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+    rules:
+
+    - rule: Rule which will wait for user message when it was applied
+      steps:
+      - intent: greet
+      - action: utter_greet
+      # Every rule implicitly includes a prediction for `action_listen` as last step.
+      # This means that Rasa Open Source will wait for the next user message.
+      - action_listen
```
Comment (Ghostvv) [inline docs/core/rules.rst:121]: this looks weird in the docs. I don't think unexperienced user will get it

File: docs/core/rules.rst [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    - rule: Rule which will not wait for user message once it was applied
+      steps:
+      - intent: greet
+      - action: utter_greet
+      - ...
+
+Rules and Forms
+~~~~~~~~~~~~~~~
+
+Rules don't apply when a :ref:`forms` is active. Rules become applicable again if
```
Comment (Ghostvv) [inline docs/core/rules.rst:140]: I think "don't apply" is too bold. Maybe
```suggestion
When a :ref:`forms` is active, rules become applicable if
```

File: docs/core/fallback-actions.rst (hunk 4 above)
Comment (wochinge) [reply docs/core/fallback-actions.rst:42]: You mean with the `condition: ` step? Yes, all the changes from the feedback has to be worked in.

File: docs/core/fallback-actions.rst (hunk 5 above)
Comment (wochinge) [reply docs/core/fallback-actions.rst:144]: No. I stated that in the PR description.

File: docs/core/forms.rst (hunk 9 above)
Comment (wochinge) [reply docs/core/forms.rst:456]: Yes, documented in the PR description

**Comment by wochinge** [comment]
@Ghostvv As multiple comments are referring to this:
The documentation contains some parts which aren't implemented yet (described in the PR description) and also not yet the latest part of the feedback as some of it is still work in progress. Despite that there are necessary implementation steps before this PR can be merged, Ben can already review the majority of it. I'll convert back to a draft to make it clearer.

File: docs/core/forms.rst (hunk 6 above)
Comment (wochinge) [reply docs/core/forms.rst:369]: fixed it 👍

File: docs/core/fallback-actions.rst (hunk 4 above)
Comment (Ghostvv) [reply docs/core/fallback-actions.rst:42]: yes, so I meant, should it be part of this PR, or part of condition PR

File: docs/core/forms.rst (hunk 8 above)
Comment (wochinge) [reply docs/core/forms.rst:377]: What do you mean with `because it repeats the default behavior` ? It's an example for requesting slots from a custom action which should only be done if having the slot mappings in the domain doesn't cut it for you.

File: docs/core/fallback-actions.rst (hunk 4 above)
Comment (wochinge) [reply docs/core/fallback-actions.rst:42]: The condition PR will definitely go in before this one, so I'd say this PR.

File: docs/core/rules.rst (hunk 11 above)
Comment (wochinge) [reply docs/core/rules.rst:72]: see my comment here https://github.com/RasaHQ/rasa/pull/6195#issuecomment-662407289

File: docs/core/forms.rst (hunk 8 above)
Comment (Ghostvv) [reply docs/core/forms.rst:377]: yes, but this particular example can be covered by putting these slots in the domain. So users who read it might think they need to do it anyhow

File: docs/core/rules.rst (hunk 12 above)
Comment (wochinge) [reply docs/core/rules.rst:121]: This will also change with the changes from the feedback

File: docs/core/forms.rst (hunk 8 above)
Comment (wochinge) [reply docs/core/forms.rst:377]: It's in the `advanced usage` section and then I say

> If you have frequent changes to the required slots and don't want to retrain your
assistant when something in your form changes, you can also use a
:ref:`custom action<custom-actions>` ``validate_{form_name}`` to define
which slots should be requested.

I think it should be pretty clear

File: docs/core/fallback-actions.rst [hunk 14]
```diff
@@ -8,87 +8,74 @@ Fallback Actions
...
-  policies:
-    - name: "FallbackPolicy"
-      nlu_threshold: 0.4
-      core_threshold: 0.3
-      ambiguity_threshold: 0.1
-      fallback_action_name: "action_default_fallback"
+    - rule: Ask the user to rephrase whenever they send a message with low NLU confidence
+      steps:
+      - ...
+      - intent: nlu_fallback
```
Comment (akelad) [inline docs/core/fallback-actions.rst:43]: @b-quachtran @wochinge does this intent need to be added to the domain file? Or do we silently add it?

File: docs/core/forms.rst (hunk 8 above)
Comment (Ghostvv) [reply docs/core/forms.rst:377]: but people might first look into example, so I suggest to create example that illustrates what is written there

File: docs/core/fallback-actions.rst (hunk 14 above)
Comment (b-quachtran) [reply docs/core/fallback-actions.rst:43]: As of right now, Rasa will throw a warning if it isn't added to the domain file.

**Comment by akelad** [comment]
@b-quachtran let me know when you need a review

**Comment by b-quachtran** [comment]
@akelad Read my mind :sweat_smile:

File: docs/core/fallback-actions.rst (hunk 14 above)
Comment (wochinge) [reply docs/core/fallback-actions.rst:42]: It's in the default intents - so it should work 🤔 What warning do you get @b-quachtran ?

**Comment by wochinge** [comment]
@b-quachtran @akelad I'm back 👋  Am I right that the next steps would be:
1. convert to new docs format (me)
2. update to latest training data format from feedback document (me)
3. review + refine (Ben, Vova)
3. merge

**Comment by wochinge** [comment]
@m-vdb Would you recommend using a tool to convert the docs or doing it manually? There are no fancy elements in there.

**Comment by m-vdb** [comment]
@wochinge any generic tool works, but might require some manual updates as well. If it's not too long to do, I'd even prefer to manually change the page (replace each titles + sphinx directive).

**Comment by m-vdb** [comment]
@wochinge also happy to handle the merge conflicts if you want (i've got a few under my belt...)

**Comment by akelad** [comment]
ok sounds like my review isn't needed after all then - let me know if you need anything though!

**Comment by wochinge** [comment]
> ok sounds like my review isn't needed after all the 

Your review is always welcome and highly appreciated!

**Comment by wochinge** [comment]
Migrated everything to 🦖  I tried to move all your copy / training data changes correctly @b-quachtran . I'll have another look if on the migration tomorrow. I need a documentation break after the migration 😄

File: docs/docs/core/fallback-actions.mdx [hunk 15]
```diff
@@ -6,117 +6,146 @@ description: Define custom fallback actions with thresholds for NLU and Core for
...
-file as parameters of the `FallbackPolicy`.
+`"Sorry, I didn't understand that"`. You can handle fallback cases by adding appropriate
+rules. Rasa Open Source comes with two default implementations for handling these
+fallbacks.
+In addition, you can also use [Custom Actions](./actions.mdx#custom-actions) to run any
+custom code.
+
+## Handling Low NLU Confidence
+
+Although Rasa's [Intent Classifier](../nlu/components.mdx#intent-classifiers) will
```
Comment (m-vdb) [inline docs/docs/core/fallback-actions.mdx:17]: I found out that relative links sometimes break, does it pass the build?

File: docs/docs/core/fallback-actions.mdx [hunk 16]
```diff
@@ -6,117 +6,146 @@ description: Define custom fallback actions with thresholds for NLU and Core for
...
       is asked to affirm the classified intent.
 
-* Second affirmation  (default action: `action_default_ask_affirmation`)
+- Second affirmation  (default action: `action_default_ask_affirmation`)
 
-    * If the user affirms the intent, the story continues as if the user had
-      this intent from the beginning.
+    - If they affirm by sending a message with high NLU confidence (e.g. by pressing
+      a button), the story continues as if the user had this intent from the beginning.
+    - If the user denies by sending a message with the intent ``out_of_scope``, the
```
Comment (m-vdb) [inline docs/docs/core/fallback-actions.mdx:85]: double backticks -> sing backtick

File: docs/docs/core/fallback-actions.mdx [hunk 17]
```diff
@@ -6,117 +6,146 @@ description: Define custom fallback actions with thresholds for NLU and Core for
...
+      original intent is classified as the specifies ``deny_suggestion_intent_name``,
```
Comment (m-vdb) [inline docs/docs/core/fallback-actions.mdx:86]: double backticks -> sing backtick

File: docs/docs/core/fallback-actions.mdx [hunk 18]
```diff
@@ -6,117 +6,146 @@ description: Define custom fallback actions with thresholds for NLU and Core for
...
+
+`action_default_fallback` is a default action in Rasa Open Source that sends the
+`utter_default` response to the user. Make sure to specify
+the `utter_default` in your domain file. It will also revert back to the
+state of the conversation before the user message that caused the
+fallback, so it will not influence the prediction of future actions.
+You can take a look at the source of the action below:
+
+TODO!!!!
+.. autoclass:: rasa.core.actions.action.ActionDefaultFallback
```
Comment (m-vdb) [inline docs/docs/core/fallback-actions.mdx:148]: [currently not implemented](https://github.com/RasaHQ/rasa/issues/6194), so you might have to find a fallback here (no pun intended!)

File: docs/docs/core/forms.mdx [hunk 19]
```diff
@@ -5,276 +5,402 @@ title: Forms
...
-  - restaurant_form
-actions:
-  ...
+policies:
+- ... # other policies
+- name: RulePolicy
 ```
 
-See `examples/formbot/domain.yml` for an example.
+<a aria-hidden="true" tabIndex="-1" className="anchor enhancedAnchor" id="form-domain"></a>
```
Comment (m-vdb) [inline docs/docs/core/forms.mdx:24]: not required, docusaurus generates them by default for every heading, I believe

File: docs/docs/core/forms.mdx [hunk 20]
```diff
@@ -5,276 +5,402 @@ title: Forms
...
 
-```story
-## happy path
-* request_restaurant
-    - restaurant_form
-    - form{"name": "restaurant_form"}
-    - form{"name": null}
+:::note
+The `form: your_form` step indicates that the form should be activated after
+``your_form`` was run.
```
Comment (m-vdb) [inline docs/docs/core/forms.mdx:69]: double backticks -> single backtick

File: docs/docs/core/forms.mdx [hunk 21]
```diff
@@ -5,276 +5,402 @@ title: Forms
...
-Once all the slots are filled, the `submit()` method is called, where you can
-use the information you’ve collected to do something for the user, for example
-querying a restaurant API.
-If you don’t want your form to do anything at the end, just use `return []`
-as your submit method.
-After the submit method is called, the form is deactivated,
-and other policies in your Core model will be used to predict the next action.
+:::note
+The slot mapping will not apply during the initial activation of the form. To fill
+a slot based on the intent that activated the form, use the ``from_trigger_intent``
```
Comment (m-vdb) [inline docs/docs/core/forms.mdx:160]: single backticks

File: docs/docs/glossary.mdx [hunk 22]
```diff
@@ -95,6 +95,12 @@ description: Glossary for all Rasa-related terms
 
   Incoming messages are processed by a sequence of components called a pipeline. A component can perform tasks ranging from entity extraction to intent classification to pre-processing.
 
+## [Rules](.core/rules.mdx)
```
Comment (m-vdb) [inline docs/docs/glossary.mdx:98]: missing slash

File: docs/docs/core/forms.mdx (hunk 19 above)
Comment (wochinge) [reply docs/docs/core/forms.mdx:24]: why is it then in all other `mdx` files? I quickly skimmed over them and adapt it from there.

Comment (m-vdb) [reply docs/docs/core/forms.mdx:24]: it came from the automatic generation.

Comment (wochinge) [reply docs/docs/core/forms.mdx:24]: ah, ok. So I can remove them all?

Comment (m-vdb) [reply docs/docs/core/forms.mdx:24]: yeah the one in your doc. But leave the other ones as it could break permalinks

File: docs/docs/core/fallback-actions.mdx (hunk 15 above)
Comment (wochinge) [reply docs/docs/core/fallback-actions.mdx:17]: yep, passes 👍 Should I change it to absolute links anyhow?

Comment (m-vdb) [reply docs/docs/core/fallback-actions.mdx:17]: no if they do not break, keep it like that. I need to figure out why it didn't work before, not even sure I can repro now

**Comment by wochinge** [comment]
@b-quachtran Can you please give this a final review so that we can get it in the next alpha?
The documentation is still ahead of some smaller code things, but that's ok imo as the major things are in. Will handle the smaller code todos in the next sprint.

**Comment by wochinge** [comment]
It's merged! 🎉 Thanks a lot for your help @b-quachtran 🙏 💯
